feat: add Gen

srghma opened the pull request.

srghma commented:
@natefaubion can I be added as maintainer to this repo?

natefaubion commented on spago.yaml:
```diff
@@ -0,0 +1,41 @@
+      - record
+      - tuples
+      - unsafe-coerce
+workspace:
+  extraPackages: {}
+  packageSet:
+    registry: 64.3.0
```
Can we remove the registry? I would prefer to just not have that in the configuration. It doesn't serve any purpose as a library that is published with version ranges.

natefaubion commented on src/Data/Variant/Gen.purs:
````diff
@@ -0,0 +1,176 @@
+--     }
+-- ```
+--
+-- Inspired by https://github.com/JordanMartinez/purescript-veither/blob/9fb195b7bd97cfcf21bbd389b55002ca458f1bb4/src/Data/Veither.purs#L461
````
What is the inspiration exactly and how is it relevant? Do we need to fulfill a license? If so, I don't think the link is enough. If there are important, non-obvious considerations, it probably warrants a general explanation comment somewhere documenting the cases with this link as a reference. Otherwise, I don't think the link is necessary since it's not obvious that it's contextually relevant to anything.

natefaubion commented on src/Data/Variant/Gen.purs:
````diff
@@ -0,0 +1,176 @@
+import Record as Record
+import Type.Proxy (Proxy(..))
+
+-- Generate a `Variant` with uniform probability given a record whose
+-- generators' labels correspond to the `Variant`'s labels
+--
+-- ```purs
+-- genMyVariant :: Gen (Variant (id :: String, unit :: Unit))
+-- genMyVariant =
+--   genVariantUniform
+--     { id: genAlphaString
+--     , unit: pure unit
+--     }
+-- ```
````
Should this use doc comment style: `-- | `?

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  -> m (Variant variantRows)
+genVariantUniform = oneOf <<< buildGenVariantUniformNonEmptyList (Proxy :: Proxy rowList)
+
+-- For what functions we can use `NonEmpty List ...` as input? For now - only for `oneOf`
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  where
+  buildGenVariantUniformNonEmptyList :: Proxy rowList -> Record genRows -> NonEmpty L.List (m (Variant variantRows))
+
+-- Instance for a list with at least one element
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+    in
+      makeVariant :| tail
+
+-- Class for handling the rest of the list
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  where
+  buildGenVariantUniformList :: Proxy rowList -> Record genRows -> L.List (m (Variant variantRows))
+
+-- Instance for the rest of the list when it has elements
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+    in
+      L.Cons makeVariant rest
+
+-- Base case for rest of list
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  GenVariantUniformList RL.Nil genRows variantRows m where
+  buildGenVariantUniformList _ _ = L.Nil
+
+------------------------------------------------------------
```
Please avoid ASCII decorations in comments.

natefaubion commented on src/Data/Variant/Gen.purs:
````diff
@@ -0,0 +1,176 @@
+  buildGenVariantUniformList _ _ = L.Nil
+
+------------------------------------------------------------
+-- Generate a `Variant` with uniform probability given a record whose
+-- generators' labels correspond to the `Variant`'s labels
+--
+-- ```purs
+-- genMyVariant :: Gen (Variant (id :: String, unit :: Unit))
+-- genMyVariant =
+--   genVariantFrequency
+--     { id: Tuple 0.1 genAlphaString
+--     , unit: Tuple 0.9 (pure unit)
+--     }
+-- ```
````
Should this use doc comment style: '-- | `?

natefaubion commented on src/Data/Variant/Gen.purs:
````diff
@@ -0,0 +1,176 @@
+--     , unit: Tuple 0.9 (pure unit)
+--     }
+-- ```
+--
+-- Inspired by https://github.com/JordanMartinez/purescript-veither/blob/9fb195b7bd97cfcf21bbd389b55002ca458f1bb4/src/Data/Veither.purs#L461
````
Same comment here about inspiration.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  -> m (Variant variantRows)
+genVariantFrequency r = frequency $ buildGenVariantFrequencyNonEmptyList (Proxy :: Proxy rowList) r
+
+-- For what functions we can use `NonEmpty List ...` as input? For now - only for `oneOf`
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  where
+  buildGenVariantFrequencyNonEmptyList :: Proxy rowList -> Record genRows -> NonEmpty L.List (Tuple Number (m (Variant variantRows)))
+
+-- Instance for a list with at least one element
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+    in
+      head :| tail
+
+-- Class for handling the rest of the list
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  where
+  buildGenVariantFrequencyList :: Proxy rowList -> Record genRows -> L.List (Tuple Number (m (Variant variantRows)))
+
+-- Instance for the rest of the list when it has elements
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+    in
+      L.Cons current rest
+
+-- Base case for rest of list
```
I don't think this comment adds anything.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  GenVariantUniformNonEmptyList (RL.Cons name (m a) restRowList) genRows variantRows m where
+  buildGenVariantUniformNonEmptyList _ rec =
+    let
+      nameProxy = Proxy :: Proxy name
+      restProxy = Proxy :: Proxy restRowList
+      (generator :: m a) = Record.get nameProxy rec
+      (makeVariant :: m (Variant variantRows)) = map (inj nameProxy) generator
```
This mixes type annotation styles. If we aren't going to double up the names, can we settle on "postfix" style like with Proxy? That's what's used everywhere else. If the non-proxy signatures are unnecessary for type-checking, maybe just omit them?

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+  GenVariantUniformList (RL.Cons name (m a) restRowList) genRows variantRows m where
+  buildGenVariantUniformList _ rec =
+    let
+      nameProxy = Proxy :: Proxy name
+      restProxy = Proxy :: Proxy restRowList
+      (generator :: m a) = Record.get nameProxy rec
+      (makeVariant :: m (Variant variantRows)) = map (inj nameProxy) generator
```
Same comment here regarding type annotations.

natefaubion commented on src/Data/Variant/Gen.purs:
```diff
@@ -0,0 +1,176 @@
+    let
+      nameProxy = Proxy :: Proxy name
+      restProxy = Proxy :: Proxy restRowList
+      (Tuple freq generator :: Tuple Number (m a)) = Record.get nameProxy rec
```
Same here regarding type annotations. If they are necessary, it seems like it would be enough to annotate only the necessary binding?

natefaubion commented:
> @natefaubion can I be added as maintainer to this repo? or lets move to contrib?

I'm not opposed to adding other maintainers, but my experience with your PRs has been somewhat mixed. I appreciate that you want to contribute, but I often find:
* Significant work done without reaching out to maintainers, offered in a large PR absent maintainer guidance.
* Incomplete or temporary work that doesn't get updated to a merge-able state.
* Additional unsolicited, opinionated changes or refactorings.
* Coding styles and conventions that don't reflect the existing codebase.

Note, I'm not saying any given change or opinion is "bad" on it's face. And certainly if you are working on your own project, your workflow is just fine. But when working on a codebase authored by someone else, I'd ask that you adhere to an etiquette:
* Keep patches focused.
* Don't do unsolicited work. Open a ticket or reach out in someway to make sure the work will be received.
* Don't do unsolicited refactorings in the process of addressing solicited work.
* Write code that looks like the existing codebase. 
* Be conservative about breaking changes.

For example, this patch doubles up adding gen and migrating to spago next, and it's not obvious to me why that is necessary. Adding gen certainly could have been done without upgrading. I'm pretty sure there is no actual dependency between the two.

I write my projects for me, and so I want my projects to look like my code and use the conventions and tooling I like to use. I enjoy them and enjoy working on them at my leisure. That is not always consistent, and I offer no guarantees and declare no obligations to maintain code that I don't actively use in some way. If that is frustrating for you, then you should fork! That's OK, and the way open source is supposed to work.

srghma commented:
> Incomplete or temporary work that doesn't get updated to a merge-able state.

`https://github.com/purescript-halogen/purescript-halogen/pull/728` ? war, my home near Pokrovsk, but I finished `https://github.com/purescript-halogen/purescript-halogen/pull/833` 

> Coding styles and conventions that don't reflect the existing codebase.

`https://github.com/purescript-halogen/purescript-halogen-vdom/pull/39` ? changed `a b c` to meaningful names


> I'd ask that you adhere to an etiquette 

split on many prs, ok. Can be a maintainer? But better move to contrib bc some contrib package wanted to use variant IIRC

natefaubion commented:
The above critique applies to a number of interactions across repos I personally maintain, where I am expected to review your work. Again, to be clear, it is not meant as a judgment, or meant to invalidate those opinions. It's an explanation as to why I don't really merge your PRs, and why I'm not very responsive to PRs in general.

> Can be a maintainer? But better move to contrib bc some contrib package wanted to use variant IIRC

Is there something you specifically need to accomplish or afundamental issue that is not being addressed?

> war, my home near Pokrovsk,

I'm very sorry to hear that, and you have my sincere sympathy.

srghma commented:
> something you specifically need to accomplish

heroui+nextjs for halogen or deku

> why I'm not very responsive to (Your) PRs in general

everyone ....... `a b c` names and no comments is bad

natefaubion commented:
> heroui+nextjs for halogen or deku

I'm specifically referring to this repo. What are you trying to accomplish as a maintainer of this repo? Specific tasks, issues, plans?

> why I'm not very responsive to (Your) PRs in general

The inserted "Your" is not correct. I truly meant "PRs in general" when referring to my personal repos I maintain.

> everyone ....... a b c names and no comments is bad

As far as I am aware, every public function is documented. You are correct that I have a personal dislike for comments that restate code superficially in prose as well as ASCII decorations. I don't think they add to understanding, and I think they grow stale quickly. I have no problem with comments that explain tricky invariants, obscure issues, or assumptions unavailable to the type system.

I understand your viewpoint, I'm just not really interested in debating this for my own projects.

srghma commented:
> accomplish as a maintainer of this repo

merge prs like change `export.foo` to `export const foo` or `spago update`, fix ci, not breaking, no publishing (unless (I dont want that) You disappear)

> restate code superficially

I was more about `https://github.com/purescript-halogen/purescript-halogen-vdom/pull/39`. I like Your style here.
